### ISSUE TITLE
[X86] Infinite loop in DAG combiner for targets with the fast-lzcnt feature

### ISSUE THREAD

**Issue by goussepi** [llvm/llvm-project#54694]
```
define i32 @testmasks(i32 %0, i32 %1) {
  %3 = and i32 %0, 131072
  %4 = icmp eq i32 %3, 0
  %5 = and i32 %1, 131072
  %6 = icmp eq i32 %5, 0
  %7 = select i1 %4, i1 true, i1 %6
  %8 = select i1 %7, i32 1, i32 2
  ret i32 %8
}
```

llc -mtriple=x86_64-pc-linux -mcpu=btver2 mask.ll -O3 -o -
        .text
        .file   "lzcnt2.ll"

**Comment by RKSimon** [comment]
Fixed by a3d5f1cf5d88dfbbed931951e07f328d5ceba510 - @goussepi should we cherry pick this for 14.x ?

**Comment by RKSimon** [comment]
(discussed this with @goussepi offline)

**Comment by tstellar** [comment]
@RKSimon @goussepi There is a test failure with this patch in the release/14.x: https://github.com/llvmbot/llvm-project/pull/159

Can someone update the patch and submit a branch.

**Comment by goussepi** [comment]
Thank you @tstellar for reporting, I am having a look at it.

**Comment by wjristow** [comment]
Discussed this with @goussepi and he asked me to take care of the branch

**Comment by tstellar** [comment]
Merged: b83c4a2dc0fb b83c4a2dc0fb
